lib/wcgi-runner/src/builder.rs

```diff
@@ -0,0 +1,252 @@
+            ..self
+        }
+    }
+
+    /// Add an argument to the WASI executable's command-line arguments.
+    pub fn arg(mut self, arg: impl Into<String>) -> Self {
+        self.args.push(arg.into());
+        self
+    }
+
+    /// Add multiple arguments to the WASI executable's command-line arguments.
+    pub fn args<A, S>(mut self, args: A) -> Self
+    where
+        A: IntoIterator<Item = S>,
+        S: Into<String>,
+    {
+        self.args.extend(args.into_iter().map(|s| s.into()));
+        self
+    }
+
+    /// Expose an environment variable to the guest.
+    pub fn env(mut self, name: impl Into<String>, value: impl Into<String>) -> Self {
+        self.env.insert(name.into(), value.into());
+        self
+    }
+
+    /// Expose multiple environment variables to the guest.
+    pub fn envs<I, K, V>(mut self, variables: I) -> Self
+    where
+        I: IntoIterator<Item = (K, V)>,
+        K: Into<String>,
+        V: Into<String>,
+    {
+        self.env
+            .extend(variables.into_iter().map(|(k, v)| (k.into(), v.into())));
+        self
+    }
+
+    /// Forward all of the host's environment variables to the guest.
+    pub fn forward_host_env(self) -> Self {
+        Builder {
+            forward_host_env: true,
+            ..self
+        }
+    }
+
+    /// Override the CGI dialect.
+    pub fn cgi_dialect(self, dialect: CgiDialect) -> Self {
+        Builder {
+            dialect: Some(dialect),
+            ..self
+        }
+    }
+
+    /// Map `guest_dir` to a directory on the host.
+    pub fn map_dir(mut self, guest_dir: impl Into<String>, host_dir: impl Into<PathBuf>) -> Self {
+        self.mapped_dirs.push((guest_dir.into(), host_dir.into()));
+        self
+    }
+
+    /// Map one or more `guest_dir` to a directory on the host.
+    pub fn map_dirs<I, G, H>(mut self, mapping: I) -> Self
+    where
+        I: IntoIterator<Item = (G, H)>,
+        G: Into<String>,
+        H: Into<PathBuf>,
+    {
+        for (guest_dir, host_dir) in mapping {
+            self.mapped_dirs.push((guest_dir.into(), host_dir.into()));
+        }
+
+        self
+    }
+
+    /// Pass in a [`Handle`] to a custom tokio runtime.
+    pub fn tokio_handle(self, handle: Handle) -> Self {
+        Builder {
+            tokio_handle: Some(handle),
+            ..self
+        }
+    }
+
+    /// Set the engine used to compile the WebAssembly module.
+    pub fn engine(self, engine: Engine) -> Self {
+        Builder {
+            engine: Some(engine),
+            ..self
+        }
+    }
+
+    /// Create a [`Runner`] that executes a WEBC file.
+    ///
+    /// If [`Builder::program`] was set, this will look for the command with
+    /// that name in the WEBC file. Otherwise, it will fall back to the WEBC
+    /// file's default entrypoint.
+    ///
+    /// This will infer the [`CgiDialect`] from the WEBC file's metadata
+    pub fn build_webc(self, webc: impl Into<Bytes>) -> Result<Runner, Error> {
```

Comment:
We face a difficult layering problem with webc.
I also had to contend with the same issue in ... another product...

I arrived at the conclusion that what we want is:

* A webc runner that understands entrypoints and commands, extracts the required configuration and forwards to a concrete runners
* Concrete runners that are not webc-aware at all

If we don't do this we need to plaster webc logic all over every runner we implement, which feels like a lot of duplication and a layering violation to me.

Reply:
Ignoring the above, I feel `new_webc` would be more appropriate here, and the method would also need to to take an`command: Option<String>`.